py/server/tests/test_table.py

```diff
@@ -1063,6 +1063,19 @@ def test_agg_count_and_partition_error(self):
             t.agg_by(aggs=[partition(["A"])], by=["B"])
         self.assertIn("string value", str(cm.exception))
 
+    def test_agg_formula_scope(self):
+        def run_func_work():
+            def my_fn(vals):
+                import deephaven.dtypes as dht
+                return dht.array(dht.double, [i + 2 for i in vals])
+
+            t = empty_table(1000).update_view(["A=i%2", "B=A+3"])
+            t = t.agg_by([formula("(double[])my_fn(each)", formula_param='each', cols=['C=B'])], by='A') # This
+            return t
+
+        t = run_func_work()
+        self.assertIsNotNone(t)
+
```

Comment:
The way this feature is implemented, I think there is a case we fail in, but maybe we don't care.

```python

def make_agg():
    def my_fn(x):
        return x

    return formula("(double[])my_fn(each)", formula_param='each', cols=['C=B'])

t = t.agg_by([make_agg()], by='A')
```
Here there is a disconnect in the `agg_by` and the `formula` scopes, which leads to problems.

Reply:
Right, only the scope in which `Table.agg_by` is called is visible to the formulas. I don't think we need to care about this case, especially when it can be easily worked around.

Reply:
This is the same as any and all formula evaluations in the engine. We substitute for scope variables at operation initialization time, not when the expression is composed.